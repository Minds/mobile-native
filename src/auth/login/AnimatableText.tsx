import React from 'react';
import * as Animatable from 'react-native-animatable';
import ThemedStyles from '../../styles/ThemedStyles';
const AnimatableText = ({ msg }) => {
  if (!msg) {
    return null;
  }
  return (
    <Animatable.Text
      animation="bounceInLeft"
      useNativeDriver
      style={styles.text}
      testID="loginMsg">
      {msg}
    </Animatable.Text>
  );
};

const styles = ThemedStyles.create({
  text: ['subTitleText', 'colorSecondaryText', 'textCenter'],
});

export default AnimatableText;
