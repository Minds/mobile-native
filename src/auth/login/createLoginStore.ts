import i18n from '../../common/services/i18n.service';
import AuthService from '../AuthService';
import { LayoutAnimation } from 'react-native';
import logService from '../../common/services/log.service';
import { showNotification } from '../../../AppMessages';

const createLoginStore = ({ props, resetRef }) => ({
  username: '',
  password: '',
  msg: '',
  language: i18n.getCurrentLocale(),
  hidePassword: true,
  inProgress: false,
  setUsername(value) {
    let username = String(value).trim();
    // check for @ char at start, remove it if it is present.
    if (username.charAt(0) === '@') {
      username = username.substring(1);
    }
    this.username = username;
  },
  setPassword(value) {
    const password = String(value).trim();
    this.password = password;
  },
  toggleHidePassword() {
    this.hidePassword = !this.hidePassword;
  },
  setInProgress(value: boolean) {
    this.inProgress = value;
  },
  initLogin() {
    this.msg = '';
    this.inProgress = true;
  },
  setError(msg: string) {
    this.msg = msg;
    showNotification(msg, 'warning', 3000, 'top');
    this.inProgress = false;
  },
  onLoginPress() {
    this.initLogin();
    // is two factor auth
    AuthService.login(this.username, this.password)
      .then(() => {
        props.onLogin && props.onLogin();
      })
      .catch(err => {
        const errJson = err.response ? err.response.data : err;
        LayoutAnimation.configureNext(LayoutAnimation.Presets.spring);
        if (
          errJson.error === 'invalid_grant' ||
          errJson.error === 'invalid_client'
        ) {
          this.setError(i18n.t('auth.invalidGrant'));
          return;
        }

        if (errJson.message.includes('user could not be found')) {
          this.setError(i18n.t('auth.loginFail'));
          return;
        }

        this.setError(errJson.message || 'Unknown error');

        logService.exception('[LoginForm]', errJson);
      });
  },
  onForgotPress() {
    resetRef.current?.show();
  },
});

export type LoginStore = ReturnType<typeof createLoginStore>;
export default createLoginStore;
