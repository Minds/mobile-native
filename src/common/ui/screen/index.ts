export * from './Screen';
export * from './ScreenHeader';
export * from './ScreenSection';
export * from './ModalFullScreen';
