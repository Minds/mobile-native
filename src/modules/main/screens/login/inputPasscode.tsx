// import React, { useState } from 'react';
// import { Icon, Input, Text } from 'components/.';
// import { Controller } from 'react-hook-form';
// import { globalStyles } from 'styles/.';
// import { FormInputType } from 'components/global.types';
// import { useTranslation } from 'containers/main/locales';
// import { ImageProps, TouchableWithoutFeedback } from 'react-native';

// interface InputPasscodeType extends FormInputType<{ passcode?: string }> {
//   passcode?: string;
// }

// export function InputPasscode(props: InputPasscodeType): JSX.Element {
//   const { t } = useTranslation();
//   const { isVisible } = useVisibilityIcon();
//   const { passcode = '', control, errors, theRef, onSubmitEditing } = props;

//   return (
//     <Controller
//       control={control}
//       defaultValue={passcode}
//       name="passcode"
//       rules={{
//         required: { value: true, message: t('Passcode is required') },
//         pattern: {
//           value: /^\d{6}$/i,
//           message: t('Passcode must have exactly 6 digits'),
//         },
//       }}
//       render={({ onChange, onBlur, value }) => (
//         <>
//           <Input
//             ref={theRef}
//             size="large"
//             status="basic"
//             style={globalStyles.marginTopBase}
//             placeholder={t('Please enter')}
//             returnKeyType={'next'}
//             label={t('Passcode')}
//             testID={'input-passcode-field'}
//             secureTextEntry={!isVisible}
//             autoFocus={false}
//             autoCorrect={false}
//             keyboardType={'number-pad'}
//             maxLength={6}
//             value={value}
//             onBlur={onBlur}
//             onChangeText={onChange}
//             onSubmitEditing={onSubmitEditing}
//           />
//           <Text
//             testID={'text-passcode-error'}
//             category={'label'}
//             status={'danger'}>
//             {errors.passcode?.message}
//           </Text>
//         </>
//       )}
//     />
//   );
// }

// type VisibleField = {
//   isVisible: boolean;
//   renderEyeIcon: (props?: Partial<ImageProps>) => React.ReactElement;
// };

// function useVisibilityIcon(): VisibleField {
//   const [isVisible, setisVisible] = useState<boolean>(false);

//   const renderEyeIcon = (props?: Partial<ImageProps>): React.ReactElement => (
//     <TouchableWithoutFeedback onPress={() => setisVisible(value => !value)}>
//       <Icon {...props} name={isVisible ? 'eye' : 'eye-off'} />
//     </TouchableWithoutFeedback>
//   );

//   return {
//     isVisible,
//     renderEyeIcon,
//   };
// }
export {};
