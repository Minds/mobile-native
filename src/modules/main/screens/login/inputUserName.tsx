// import React from 'react'
// import { Input, Text } from 'components/.'
// import { Controller } from 'react-hook-form'
// import { globalStyles } from 'styles/.'
// import { FormInputType } from 'components/global.types'
// import { useTranslation } from 'containers/main/locales'

// interface InputUsernameType extends FormInputType<{ username?: string }> {
//   username?: string
// }

// export function InputUsername(props: InputUsernameType): JSX.Element {
//   const { t } = useTranslation()
//   const { username = '', control, errors, theRef, nextRef, onSubmitEditing } = props

//   const onSubmit = onSubmitEditing ? onSubmitEditing : () => nextRef?.current?.focus()

//   return (
//     <Controller
//       control={control}
//       defaultValue={username}
//       name='username'
//       rules={{
//         required: { value: true, message: t('This field is required') },
//         minLength: {
//           value: 4,
//           message: t('Username must have at least 4 characters'),
//         },
//       }}
//       render={({ onChange, onBlur, value }) => (
//         <>
//           <Input
//             size='large'
//             status='basic'
//             ref={theRef}
//             style={globalStyles.marginTopBase}
//             label={t('Username')}
//             testID='input-name-field'
//             placeholder={t('Please enter')}
//             returnKeyType={'next'}
//             autoFocus
//             autoCapitalize={'none'}
//             autoCorrect={false}
//             value={value}
//             onBlur={onBlur}
//             onChangeText={onChange}
//             onSubmitEditing={onSubmit}
//           />
//           <Text category={'label'} status={'danger'}>
//             {errors.username?.message}
//           </Text>
//         </>
//       )}
//     />
//   )
// }
export {};
