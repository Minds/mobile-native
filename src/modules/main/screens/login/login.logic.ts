import React from 'react';
import { ApiConnector } from 'services/api';
import { WrapperHandle } from '~/components';
import { login, registeredDevice } from '../../api';
import { useRef, useEffect, useState, useCallback } from 'react';
// import appStorage from 'services/appStorage';
// import { uuid } from 'utils/.';
import { useFocusEffect } from '@react-navigation/native';

export type BasicLoginParams = {
  username: string;
  passcode: string;
  deviceKey: string;
};

export type VerifyOtpParams = {
  otp: string;
  verificationId: string;
};

type DeviceLoginHook = {
  deviceLogin: (params: BasicLoginParams) => Promise<boolean>;
};

type RegisterDeviceHook = {
  checkDeviceAvailability: (params: BasicLoginParams) => Promise<boolean>;
  wrapperHandle: React.RefObject<WrapperHandle>;
};

type GetCredentialsHook = {
  deviceKey: string;
  partyKey: string | null;
  isNewDevice: boolean;
};

type LoginData = {
  partyKey: string;
  availableSubscriptions: Array<{ subscriptionKey: string }>;
};

export function useDeviceLogin(): DeviceLoginHook {
  const deviceLogin = async (params: BasicLoginParams) => {
    const { username, passcode, deviceKey } = params;
    ApiConnector.getInstance().updateHeaders({ 'X-DeviceKey': deviceKey });
    const { partyKey = '', availableSubscriptions = [] } =
      ((await login({ username, passcode })) as LoginData) || {};
    if (partyKey && availableSubscriptions.length > 0) {
      return true;
    }
    return false;
  };
  return { deviceLogin };
}

export function useRegisterDevice(): RegisterDeviceHook {
  const wrapperHandle = useRef<WrapperHandle>(null);
  const checkDeviceAvailability = async (params: BasicLoginParams) => {
    const { username, passcode, deviceKey } = params;
    wrapperHandle?.current?.loadingStart();
    ApiConnector.getInstance().updateHeaders({ 'X-DeviceKey': deviceKey });
    const response = (await registeredDevice({ username, passcode })) || false;
    wrapperHandle?.current?.loadingStop();
    return response;
  };
  return { checkDeviceAvailability, wrapperHandle };
}

export function useGetCredentials(reset?: () => void): GetCredentialsHook {
  const [deviceKey, setDeviceKey] = useState<string>('');
  const [partyKey, setPartyKey] = useState<string | null>(null);
  const [isNewDevice, setIsNewDevice] = useState<boolean>(false);

  const requestPartyKey = async () => {
    const partyKey = '';
    setIsNewDevice(!partyKey);
    setPartyKey(partyKey);
  };

  const requestDeviceKey = useCallback(async () => {
    setDeviceKey('');
  }, []);

  useEffect(() => {
    requestDeviceKey();
    requestPartyKey();
  }, [requestDeviceKey, reset]);

  useFocusEffect(
    useCallback(() => {
      requestDeviceKey();
      requestPartyKey();
    }, [requestDeviceKey]),
  );

  return { deviceKey, partyKey, isNewDevice };
}
