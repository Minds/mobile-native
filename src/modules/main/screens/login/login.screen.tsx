import { useNavigation } from '@react-navigation/native';
import React from 'react';
import { View, Text, Button } from 'react-native';
import { useTranslation } from '../../locales';
import {
  useDeviceLogin,
  // useRegisterDevice,
  // useGetCredentials,
} from './login.logic';

// type LoginType = {
//   username: string;
//   passcode: string;
// };
export function LoginScreen(): React.ReactElement {
  // const usernameRef = useRef(null);
  // const passcodeRef = useRef(null);
  // const formProps = useForm<LoginType>({
  //   mode: 'onChange',
  // });
  // const { formState, handleSubmit } = formProps;

  const navigation = useNavigation();
  // const modal = useAlertModal();
  const { t } = useTranslation();
  const { deviceLogin } = useDeviceLogin();
  // const { checkDeviceAvailability } = useRegisterDevice();

  // const { deviceKey, partyKey, isNewDevice } = useGetCredentials();

  /**
   * Replace below for fast login for debugging
   */

  React.useEffect(() => {
    // deviceLogin(findUserByName('Paul'));
  }, [deviceLogin]);

  const onReset = () => navigation.navigate('Reset');

  // const onLogin = async ({ username, passcode }: LoginType) => {
  //   if (!isNewDevice) {
  //     const isSuccessfulLogin = await deviceLogin({
  //       username: partyKey ? partyKey : '',
  //       passcode,
  //       deviceKey,
  //     });

  //     if (isSuccessfulLogin) {
  //       // await appStorage.setLoginUser(partyKey ? partyKey : '', passcode);
  //     }

  //     return isSuccessfulLogin;
  //   }
  //   const isValid = await checkDeviceAvailability({
  //     username,
  //     passcode,
  //     deviceKey,
  //   });
  //   if (!isValid) {
  //     // modal.show({
  //     //   title: t('Oh no'),
  //     //   message: t('Error - wrong credentials'),
  //     // });
  //     return;
  //   }
  //   navigation.navigate('RegisterDevice');
  // };

  return (
    <View>
      <Text testID="sign-in-text">{t('Sign in')}</Text>
      <View />
      {/* {isNewDevice && (
        <InputUsername
          {...{ ...formProps, theRef: usernameRef, nextRef: passcodeRef }}
        />
      )} */}
      {/* <InputPasscode
        {...{
          ...formProps,
          theRef: passcodeRef,
          onSubmitEditing: handleSubmit(onLogin),
        }}
      /> */}
      <View style={[]}>
        <Text>{t('Forgotten details?')} </Text>
        <Button title="" onPress={onReset}>
          {t('Reset')}
        </Button>
      </View>
    </View>
  );
}
