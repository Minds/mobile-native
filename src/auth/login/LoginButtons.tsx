import React from 'react';
import { View, Dimensions } from 'react-native';
import Button from '../../common/components/Button';
import MText from '../../common/components/MText';
import i18n from '../../common/services/i18n.service';
import ThemedStyles from '../../styles/ThemedStyles';
import { LoginStore } from './createLoginStore';

type PropsType = {
  localStore: LoginStore;
  multiUser?: boolean;
  onRegisterPress?: () => void;
  relogin?: boolean;
};

const { height } = Dimensions.get('window');

const LoginButtons = ({
  localStore,
  multiUser,
  onRegisterPress,
  relogin,
}: PropsType) => {
  const theme = ThemedStyles.style;
  const alt = !!(relogin || multiUser);
  return (
    <View style={styles.container}>
      <Button
        onPress={localStore.onLoginPress}
        text={i18n.t('auth.login')}
        containerStyle={styles.buttonLoginContainer}
        loading={localStore.inProgress}
        disabled={localStore.inProgress}
        testID="loginButton"
        transparent
        large
      />
      <View style={theme.marginTop4x}>
        <MText
          style={alt ? styles.forgotTextAlt : styles.forgotText}
          onPress={localStore.onForgotPress}>
          {i18n.t('auth.forgot')}
        </MText>
      </View>
      <View style={theme.flexContainer} />
      {!alt && (
        <Button
          onPress={onRegisterPress}
          text={i18n.t('auth.createChannel')}
          containerStyle={styles.buttonRegisterContainer}
          disabled={localStore.inProgress}
          testID="registerButton"
          transparent
          large
        />
      )}
    </View>
  );
};

const styles = ThemedStyles.create({
  container: ['margin6x', 'flexContainer'],
  buttonLoginContainer: ['fullWidth', { marginTop: height / 55 }],
  forgotText: ['colorWhite', 'fontL', 'textCenter'],
  forgotTextAlt: ['colorPrimaryText', 'fontL', 'textCenter'],
  buttonRegisterContainer: ['fullWidth', 'marginTop6x'],
});

export default LoginButtons;
