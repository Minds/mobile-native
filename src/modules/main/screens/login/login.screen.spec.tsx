import React from 'react';
import {
  cleanup,
  render,
  RenderAPI,
  waitFor,
} from '@testing-library/react-native';
import { LoginScreen } from './login.screen';
import { SafeAreaProvider } from 'react-native-safe-area-context';

const mockedCheckDeviceAvailability = true;
// const mockSetLoginUser = jest.fn();
const mockDeviceKey = '1234';
const mockPartyKey = '12244';
const mockIsNewDevice = false;

jest.mock('./login.logic', () => ({
  useDeviceLogin: function () {
    return {
      deviceLogin: jest.fn(),
    };
  },
  useRegisterDevice: function () {
    return {
      checkDeviceAvailability: mockedCheckDeviceAvailability,
    };
  },
  useGetCredentials: function () {
    return {
      deviceKey: mockDeviceKey,
      partyKey: mockPartyKey,
      isNewDevic: mockIsNewDevice,
    };
  },
}));

// jest.mock('services/appStorage', () => ({
//   appStorage: function () {
//     return {
//       setLoginUser: mockSetLoginUser,
//     };
//   },
// }));
describe('Login page', () => {
  let component: RenderAPI;

  beforeEach(async () => {
    component = await waitFor(() =>
      render(
        <SafeAreaProvider
          initialSafeAreaInsets={{ top: 0, left: 0, right: 0, bottom: 0 }}>
          <LoginScreen />
        </SafeAreaProvider>,
      ),
    );
  });

  afterEach(() => {
    cleanup();
  });
  it('Login snapshot test', () => {
    expect(component.toJSON).toMatchSnapshot();
  });

  it('renders the Login title correctly ', async () => {
    const { getByTestId } = component;
    expect(getByTestId('sign-in-text')).toBeTruthy();
  });
});
