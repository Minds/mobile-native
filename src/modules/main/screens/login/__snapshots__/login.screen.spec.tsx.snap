// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page Login snapshot test 1`] = `[Function]`;
